cpp/src/parquet/variant.cc

```diff
@@ -0,0 +1,744 @@
+         info.offset_size);
+  memcpy(&next_offset,
+         value.data() + info.offset_start_offset + (variantId + 1) * info.offset_size,
+         info.offset_size);
+  offset = arrow::bit_util::FromLittleEndian(offset);
+
+  if (info.data_start_offset + offset > value.size()) {
+    throw ParquetException("Invalid object field offsets: data_start_offset=" +
+                           std::to_string(info.data_start_offset) +
+                           ", offset=" + std::to_string(offset) +
+                           ", value_size=" + std::to_string(value.size()));
+  }
+
+  // Create a VariantValue for the field
+  VariantValue field_value{
+      .metadata = metadata,
+      .value = value.substr(info.data_start_offset + offset)};
+
+  return field_value;
+}
+
+VariantValue::ArrayInfo VariantValue::getArrayInfo() const {
+  checkBasicType(VariantBasicType::Array);
+  uint8_t value_header = value[0] >> 2;
+  uint8_t field_offset_size = (value_header & 0b11) + 1;
+  bool is_large = ((value_header >> 2) & 0b1);
+
+  // check the array header
+  uint8_t num_elements_size = is_large ? 4 : 1;
+  if (value.size() < 1 + num_elements_size) {
+    throw ParquetException(
+        "Invalid array value: too short: " + std::to_string(value.size()) +
+        " for at least " + std::to_string(1 + num_elements_size));
+  }
+
+  // parse num_elements
+  uint32_t num_elements = 0;
+  {
+    memcpy(&num_elements, value.data() + 1, num_elements_size);
+    num_elements = arrow::bit_util::FromLittleEndian(num_elements);
+  }
+
+  ArrayInfo info{};
+  info.num_elements = num_elements;
+  info.offset_size = field_offset_size;
+  info.offset_start_offset = 1 + num_elements_size;
+  info.data_start_offset =
+      info.offset_start_offset + (num_elements + 1) * field_offset_size;
+
+  // Boundary check
+  if (info.data_start_offset > value.size()) {
+    throw ParquetException("Invalid array value: data_start_offset=" +
+                           std::to_string(info.data_start_offset) +
+                           ", value_size=" + std::to_string(value.size()));
+  }
+
+  // Validate final offset is equal to the size of the value,
+  // it would work since even empty array would have an offset of 0.
+  {
+    uint32_t final_offset = 0;
+    memcpy(&final_offset,
+           value.data() + info.offset_start_offset + num_elements * field_offset_size,
+           field_offset_size);
+    final_offset = arrow::bit_util::FromLittleEndian(final_offset);
+
+    if (info.data_start_offset + final_offset > value.size()) {
+      throw ParquetException(
+          "Invalid array value: final_offset=" + std::to_string(final_offset) +
+          ", data_start_offset=" + std::to_string(info.data_start_offset) +
+          ", value_size=" + std::to_string(value.size()));
+    }
+  }
+
+  // checking the element is incremental.
+  // TODO(mwish): Remove this or encapsulate this range check to function
```

Comment:
I think should we use a extra function here like "Validate", or just checks them here?